cluster_experiments/experiment_analysis.py

```diff
@@ -680,3 +682,255 @@ def analysis_point_estimate(self, df: pd.DataFrame, verbose: bool = False) -> fl
+
+        Returns:
+            tuple: A tuple containing the average treatment effect and a list of the placebo effects.
+        """
+        synthetic_donors = [
+            unit.assign(effect=unit[self.target_col] - unit["synthetic"])
+            for unit in synthetic_donors
+        ]
+
+        avg_effects = [
+            unit.query(f"{self.time_col} >= '{self.intervention_date}'")[
+                "effect"
+            ].mean()
+            for unit in synthetic_donors
+        ]
+
+        treatment_index = np.where(
+            [df[self.treatment_col].any() for df in synthetic_donors]
+        )[0][0]
+        ate = avg_effects[treatment_index]  # this is the real treatment effect
+        avg_effects.pop(treatment_index)  # this becomes the placebo effects
+        return ate, avg_effects
+
+    def pvalue_based_on_hypothesis(self, synthetic_donors: List[pd.DataFrame]) -> float:
+        """
+        Returns the p-value of the analysis.
+        1. calculate the average effect after intervention for each unit.
+        2. count how many times the average effect is greater than the real treatment unit
+        3. Divide by the number of units. The result is the p-value using Fisher permutation test.
+        """
+
+        ate, avg_effects = self.calculate_effects(synthetic_donors)
+
+        if HypothesisEntries(self.hypothesis) == HypothesisEntries.LESS:
+            return np.mean(avg_effects < ate)
+        if HypothesisEntries(self.hypothesis) == HypothesisEntries.GREATER:
+            return np.mean(avg_effects > ate)
+        if HypothesisEntries(self.hypothesis) == HypothesisEntries.TWO_SIDED:
+            avg_effects = np.abs(avg_effects)
+            return np.mean(avg_effects > ate)
+        else:
+            raise ValueError(f"{self.hypothesis} is not a valid HypothesisEntries")
+
+    def analysis_pvalue(self, df: pd.DataFrame, verbose: bool = False) -> float:
+        """Returns the p-value of the analysis
+        Arguments:
+            df: dataframe containing the data to analyze
+            verbose (Optional): bool, prints the regression summary if True
+        """
+        if len(self.cluster_cols) > 1:
+            raise ValueError("Currently, only one cluster column is supported")
+
+        pre_experiment_df = df.query(f"{self.time_col} < '{self.intervention_date}'")
+        df = df.query(f"{self.time_col} >= '{self.intervention_date}'")
+
+        clusters = df[self.cluster_cols].iloc[:, 0].unique()
+
+        synthetic_donors = [
+            self.synthetic_control(
+                treatment_cluster=cluster,
+                df=df,
+                pre_experiment_df=pre_experiment_df,
+                verbose=verbose,
+            )
+            for cluster in clusters
+        ]
+
+        p_value = self.pvalue_based_on_hypothesis(synthetic_donors=synthetic_donors)
+        return p_value
+
+    def analysis_point_estimate(self, df: pd.DataFrame, verbose: bool = False) -> float:
```

Comment:
I didn't reuse it because I would have to re declare the signature of analysis_point_estimate (so it accepts a treatment_cluster param). But I think it will indeed streamline the code